### ISSUE TITLE
PHP: Only variables should be passed by reference

### ISSUE THREAD

**Issue by NicklasWallgren** [google/protobuf#2409]
```
namespace Google\Protobuf\Internal;

class Message {
...
$this->$setter(new RepeatedField($field->getType(), $field->getMessageType()->getClass()));
...
}
```

```
namespace Google\Protobuf\Internal;

class FileDescriptorSet extends \Google\Protobuf\Internal\Message {
...
   public function setFile(&$var)
   {
        GPBUtil::checkRepeatedField($var, GPBType::MESSAGE, \Google\Protobuf\Internal\FileDescriptorProto::class);
        $this->file = $var;
        $this->has_file = true;
    }
...
}
```

**Comment by TeBoring** [comment]
Could you be more specific about the problem?

**Comment by NicklasWallgren** [comment]
The "setFile" method expects its parameter to be a reference, and you cannot reference something that is not a variable.

http://php.net/manual/en/language.references.pass.php

"No other expressions should be passed by reference, as the result is undefined."

```
$this->setField(new RepeatedField....
```

**Comment by TeBoring** [comment]
But setFile expect to take a message instance, which should be a variable. Isn't it?

**Comment by NicklasWallgren** [comment]
Hmm, I must be running an old version. My bad.

New version
```
 $map_field = new MapField(
                            $key_field->getType(),
                            $value_field->getType(),
                            $value_field->getMessageType()->getClass());
                        $this->$setter($map_field);
```

Old version
```
 $this->$setter(
                            new RepeatedField(
                                $field->getType(),
                                $field->getMessageType()->getClass()));
```

**Comment by TeBoring** [comment]
Sorry, I still don't get what's wrong here.
According to the manual, following can be passed by reference:

1. Variables, i.e. foo($a)
2. New statements, i.e. foo(new foobar())
3. References returned from functions, i.e.

The example you've shown here is case 2.

**Comment by NicklasWallgren** [comment]
This issue has been resolved as of v3.1.0-alpha-1.

Example demonstrating the old issue. 
The execution will result in a error notice. "Notice: Only variables should be passed by reference"

```
class Example {}

function method(&$arg1) {}

method(new Example());
```

**Comment by TeBoring** [comment]
Thanks for investigating.
